### PR TITLE
try to fix #1494

### LINKED ISSUE #1494: Smarter link-with-namespace filter
- Wiktionary page: https://fr.wiktionary.org/wiki/halq%E2%80%99em%C3%A9ylem

Wikicode:
```
[[Stó:lō]]
```

Output:
```
None
```

Expected:
```
Stó:lō
```

The wikicode is stripped at https://github.com/BoboTiG/ebook-reader-dict/blob/fdca309f31290fdee99fe76efb2b2f33b45112c1/wikidict/utils.py#L391

Not sure what to do for now, just reporting the issue.

### PR DESCRIPTION
**Pull request by lasconic**
Fixes #1494

### CODE REVIEW

File: scripts/all-namespaces.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+for locale in locales:
+    result_array: List[str] = []
+    langurl = url.format(locale)
+    r = requests.get(langurl)
+    json = r.json()
+    namespaces = json["query"]["namespaces"]
+    for key, namespace in namespaces.items():
+        if n := namespace["*"]:
+            result_array.append(n)
+    results[locale] = result_array
```
Comment (BoboTiG) [inline scripts/all-namespaces.py:19]:
```suggestion
    results[locale] = sorted(result_array)
```

File: scripts/all-namespaces.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from typing import Dict, List
+
+import requests
+
+url = "https://{0}.wiktionary.org/w/api.php?action=query&meta=siteinfo&siprop=namespaces&format=json"
+
+results: Dict[str, List[str]] = {}
+locales = ("ca", "de", "el", "en", "es", "fr", "it", "no", "pt", "ru", "sv")
```
[the remark this answers is not in the document]
Comment (BoboTiG) [reply scripts/all-namespaces.py:8]: Meh, let's not mess with the import.

File: wikidict/utils.py [hunk 3]
```diff
@@ -288,20 +288,22 @@ def clean(text: str) -> str:
...
-        >>> clean("[[Catégorie:Localités d’Afrique du Sud en français]]")
+        >>> clean("[[Catégorie:Localités d’Afrique du Sud en français]]", locale="fr")
         ''
-        >>> clean("[[Archivo:Striped_Woodpecker.jpg|thumb|[1] macho.]]")
+        >>> clean("[[Archivo:Striped_Woodpecker.jpg|thumb|[1] macho.]]", locale="es")
         ''
-        >>> clean("[[Archivo:Mezquita de Córdoba - Celosía 006.JPG|thumb|[1]]]")
+        >>> clean("[[Archivo:Mezquita de Córdoba - Celosía 006.JPG|thumb|[1]]]", locale="es")
         ''
+        >>> clean("[[Stó:lō]]", "fr")
```
Comment (BoboTiG) [inline wikidict/utils.py:305]:
```suggestion
        >>> clean("[[Stó:lō]]", locale="fr")
```

File: wikidict/utils.py [hunk 4]
```diff
@@ -478,7 +480,7 @@ def process_templates(
     sub = re.sub
 
     # Clean-up the code
-    text = callback(wikicode)
+    text = callback(wikicode, locale=locale)
```
Comment (BoboTiG) [inline wikidict/utils.py:483]:
```suggestion
    text = callback(wikicode, locale)
```

**Comment by BoboTiG** [comment]
good for review, it's done on my side.

File: wikidict/utils.py [hunk 5]
```diff
@@ -373,13 +375,24 @@ def clean(text: str) -> str:
...
-    text = sub(rf"\[\[(?:{pattern}):.+?(?=\]\])\]\]*", "", text)
-
-    # More local links
-    text = sub(r"\[\[({{[^}]+}})\]\]", "\\1", text)  # [[{{a|b}}]] -> {{a|b}}
-    text = sub(r"\[\[[^|]+\|(.+?(?=\]\]))\]\]", "\\1", text)  # [[a|b]] -> b
+    # Namespaces
+    # [[File:...|...]] -> ''
+    # except [[File:...|{{...}}]] that will end on '{{...}}'
+    pattern = "|".join(iter(namespaces[locale]))
+    text = sub(rf"\[\[(?:{pattern}):[^\{{]+?(?=\]\])\]\]*", "", text)
```
Comment (BoboTiG) [inline wikidict/utils.py:382]: This is a ugly hack for "Annexe" and "Fichier". We want to keep namespaces using at least one template, like `[[Annexe:blablabla|{{foo:|bar}}]]` but bot simple calls like `[[Fichier:blablabla|foo|bar]]`.

It's fragile. If you can find another way, I'm open :)

Comment (BoboTiG) [reply wikidict/utils.py:382]: Maybe I found another way.

If we decided to parse those wikicode, and filter out later.
We really need to keep namespaces that doesn't ends with a filename: we could do that check. If it seems to end with a file extension, then fileter out. Else, keep the wikicode.

Comment (BoboTiG) [reply wikidict/utils.py:382]: We need to add more examples to be sure about a strategy.

**Comment by lasconic** [comment]
Sure, let's merge and see.

**Pull request merged**